### ISSUE TITLE
[BUG] Tracking server returns HTTP 500 instead of 404 for a missing artifact, causing client retry storms (esp. runs:/<run_id>/<model> for MLflow-3 logged models)

### ISSUE THREAD

**Issue by parce-kabilio** [mlflow/mlflow#24535]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Remote (self-hosted tracking server with proxied artifact access, S3 artifact store)

### MLflow version

- Client: 3.14.0
- Tracking server: 3.12.0

### System information

- Tracking server launched with `--serve-artifacts` (proxied artifact access)
- Artifact store: Amazon S3
- Python: 3.13 (client)

### Describe the problem

When the artifact-proxy download endpoint is asked for an artifact **that does not exist in the backing store**, the tracking server returns **HTTP 500 (`INTERNAL_ERROR`)** instead of **HTTP 404 (`RESOURCE_DOES_NOT_EXIST`)**.

Server-side log for the failing request (S3 backend):

```
ERROR mlflow.server.handlers: Error in _download_artifact: The following failures
occurred while downloading one or more artifacts from s3://<artifact-bucket>:
##### File <experiment_id>/<run_id>/artifacts/model #####
An error occurred (404) when calling the HeadObject operation: Not Found.
"GET /api/2.0/mlflow-artifacts/artifacts/<experiment_id>/<run_id>/artifacts/model HTTP/1.1" 500 Internal Server Error
```

This 404→500 conversion is harmful because MLflow's own HTTP client **retries 500 but not 404**. The client retry set is `{408, 429, 500, 502, 503, 504}` (`mlflow.utils.request_utils._TRANSIENT_FAILURE_RESPONSE_CODES`). So a "not found" that should fail fast instead triggers the full retry ladder (7 attempts with exponential backoff → several minutes) before the client gives up on that path.

**Where this really bites: loading an MLflow-3 logged model via `runs:/<run_id>/<model_name>`.**

In MLflow 3.x, `log_model(name="model")` stores the model as a first-class *logged model* under `<experiment_id>/models/m-<model_id>/artifacts`, **not** under the run's own artifact tree (`<experiment_id>/<run_id>/artifacts/model` never exists). When you load with `runs:/<run_id>/model`, `RunsArtifactRepository.download_artifacts` **intentionally probes the run-scoped path first**, then falls back to the associated logged model (`_download_model_artifacts` → `models:/m-<model_id>`). That fallback is correct and the load ultimately succeeds — but because the run-scoped probe returns **500 instead of 404**, the client burns the entire retry ladder on the guaranteed-miss probe before falling back. Measured: **~6 minutes per model** just in retries, versus ~1 second if the probe returned a clean 404.

If the missing-artifact response were a 404, the client would not retry, the fallback would proceed immediately, and `runs:/<run_id>/<model_name>` would be fast again.

This is the mechanism behind the "abnormally long load" reported in #16979. That issue was closed with a client-side workaround (use the logged-model URI) and labeled `area/docs`, but the underlying server-side 404→500 conversion — the actual cause of the retries — was never addressed. The workaround (loading via `models:/<model_id>` / the logged model's `artifact_location`) does avoid the problem, but any code path that legitimately requests a missing artifact still gets a 500 + retry storm rather than a fast 404.

### Root cause (traced in 3.14.0 source; the relevant code is unchanged since 3.12)

1. `mlflow/store/artifact/s3_artifact_repo.py` — `S3ArtifactRepository._download_file` calls `s3_client.download_file(...)` directly. A missing key raises a raw botocore `ClientError` (HTTP 404), **not** an `MlflowException(error_code=RESOURCE_DOES_NOT_EXIST)`.
2. `mlflow/store/artifact/artifact_repo.py` — the parallel-download aggregation collects error codes only from `MlflowException`s whose code is not `INTERNAL_ERROR`; if exactly one distinct code remains it is used, otherwise it defaults to `INTERNAL_ERROR`:
   ```python
   error_codes = {
       e.error_code
       for e in ...
       if isinstance(e, MlflowException) and e.error_code != "INTERNAL_ERROR"
   }
   error_code = (
       ErrorCode.Value(error_codes.pop()) if len(error_codes) == 1 else INTERNAL_ERROR
   )
   raise MlflowException("The following failures occurred while downloading ...", error_code=error_code)
   ```
   Since the botocore error is not an `MlflowException`, `error_codes` is empty → falls through to `INTERNAL_ERROR` → HTTP 500.
3. `mlflow/server/handlers.py` — `_download_artifact` is wrapped by `catch_mlflow_exception`, which sets the HTTP status from `MlflowException.get_http_status_code()`; `INTERNAL_ERROR` → 500.

### What to reproduce it

1. Start a tracking server with `--serve-artifacts` and an S3 artifact store.
2. `GET /api/2.0/mlflow-artifacts/artifacts/<any/nonexistent/path>` → observe **500** (expected: 404).
3. Or, end to end: `mlflow.sklearn.log_model(model, name="model")` in a run, then `mlflow.sklearn.load_model(f"runs:/{run_id}/model")` with client debug logging on → observe repeated `500` retries against `.../<run_id>/artifacts/model` for minutes before the load succeeds via the logged-model fallback.

### Expected behavior

A request for an artifact that does not exist in the backing store should return **HTTP 404 (`RESOURCE_DOES_NOT_EXIST`)**, so the client does not treat it as a transient failure and retry. Suggested fix: translate the backend "object not found" (e.g. botocore `ClientError` with 404 / `NoSuchKey`) into `MlflowException(..., error_code=RESOURCE_DOES_NOT_EXIST)` in the artifact repositories (at least `S3ArtifactRepository`), so the aggregation and the server handler surface a 404.

### Code to reproduce (client-side symptom)

```python
import logging
import mlflow, mlflow.sklearn
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import load_iris

logging.basicConfig(level=logging.DEBUG)
mlflow.set_tracking_uri("http://<tracking-server>:5000")
mlflow.set_experiment("repro")

X, y = load_iris(return_X_y=True)
with mlflow.start_run() as run:
    mlflow.sklearn.log_model(RandomForestClassifier().fit(X, y), name="model")
    run_id = run.info.run_id

# Retries 500 against .../<run_id>/artifacts/model for minutes before the
# logged-model fallback succeeds.
mlflow.sklearn.load_model(f"runs:/{run_id}/model")
```

### Willing to submit a PR?

- [ ] Yes, I'd be willing to submit a PR to fix this issue.

**Comment by B-Step62** [comment]
@parce-kabilio Thanks for the report! Yeah s3 artifact repo should propagate the underlying botocore error code. Would you be interested in contributing for the fix?

**Comment by parce-kabilio** [comment]
@B-Step62 yeah, I have a branch ready for PR when you say the word. While implementing the change, Claude found that the exact same gap is present when using GCS for the storage backend. Do you want me to open a second PR with the GCS fix?

**Comment by sanskar-singh-2403** [comment]
hey @B-Step62 @parce-kabilio  can i take this up?

**Comment by parce-kabilio** [comment]
@sanskar-singh-2403 Okay for me, although I had the branch ready for PR, and was waiting on @B-Step62's green light.

**Comment by UgaTheDev** [comment]
I looked into this and can confirm the root cause.

`S3ArtifactRepository._download_file` calls `s3_client.download_file(...)` with no error handling, so a botocore `ClientError` propagates raw. `ArtifactRepository.download_artifacts` has no way to classify a provider-specific exception, so it reports the failure as `INTERNAL_ERROR` → **HTTP 500**. The MLflow HTTP client retries 5xx but not 404, so a lookup that is a guaranteed miss burns the entire retry ladder instead of failing fast. That's the retry storm described here, and it's worst for `runs:/<run_id>/<model>` probing in MLflow 3, where missing-artifact lookups are a normal control-flow path rather than an error.

Worth noting that the sibling method `list_artifacts` in the same class already does this mapping, via the module-level `BOTO_TO_MLFLOW_ERROR` table — `_download_file` is simply missing it. So the fix reuses an existing table and an existing pattern rather than introducing anything new; a missing key then surfaces `RESOURCE_DOES_NOT_EXIST` → HTTP 404.

I have a tested fix with regression tests — parametrized over the botocore error codes (including the bare `403`/`404` that `HeadObject` returns, since those responses carry no error body) plus an end-to-end case asserting the HTTP status through the public `download_artifacts` entry point. Verified failing before and passing after, and I'm happy to open a PR.

One process question before I do: #24658, #24708, #25093 and #25095 were all auto-closed by the triage bot for this issue missing the `ready` label. Would a maintainer be able to triage this one so a new PR doesn't meet the same fate? Happy to wait until then.

(Scope note: this fix is limited to the S3 repository, matching the issue. The GCS and Azure `_download_file` implementations have the same gap, but each needs its own SDK-specific mapping, so those seem better as separate changes — happy to follow up if useful.)

**Comment by anshusaurav** [comment]
I'd like to work on this. The tracking server should return 404 instead of 500 when an artifact path doesn't exist. I'll handle the missing-key / not-found cases in the artifact repository backends (S3, GCS, local) and return the appropriate HTTP status.
